bin/propolis-server/src/lib/stats/virtual_machine.rs

```diff
@@ -0,0 +1,526 @@
+        kstat.ks_module == VMM_KSTAT_MODULE_NAME
+    }
+
+    fn to_samples(
+        &self,
+        kstats: KstatList<'_, '_>,
+    ) -> Result<Vec<Sample>, Error> {
+        // First, we need to map the instance's control plane UUID to the kstat
+        // instance. We'll find this through the `vmm:<instance>:vm:vm_name`
+        // kstat, which lists the instance's UUID as its name. The
+        // `VirtualMachine` target stores that internally as the `vm_name`
+        // field.
+        //
+        // Note that if this code is run from within a Propolis zone, there is
+        // exactly one `vmm` kstat instance in any case.
+        let instance = kstats
+            .iter()
+            .find_map(|(_, kstat, data)| {
+                kstat_instance_from_instance_id(kstat, data, &self.vm_name)
+            })
+            .ok_or_else(|| Error::NoSuchKstat)?;
+
+        // Armed with the kstat instance, find all relevant metrics related to
+        // this particular VM. For now, we produce only vCPU usage metrics, but
+        // others may be chained in the future.
+        let vcpu_stats = kstats.iter().filter(|(_, kstat, _)| {
+            // Filter out those that don't match our kstat instance.
+            if kstat.ks_instance != instance {
+                return false;
+            }
+
+            // Filter out those which are neither a vCPU stat of any kind, nor
+            // for one of the vCPU IDs we know to be active.
+            let Some(suffix) = kstat.ks_name.strip_prefix(VCPU_KSTAT_PREFIX)
+            else {
+                return false;
+            };
+            let Ok(vcpu_id) = suffix.parse::<u32>() else {
+                return false;
+            };
+            vcpu_id < self.n_vcpus
+        });
+        produce_vcpu_usage(self, vcpu_stats)
+    }
+}
+
+// Given a kstat and an instance's ID, return the kstat instance if it matches.
+fn kstat_instance_from_instance_id(
+    kstat: &Kstat<'_>,
+    data: &Data<'_>,
+    instance_id: &str,
+) -> Option<i32> {
+    // Filter out anything that's not a `vmm:<instance>:vm` named kstat.
+    if kstat.ks_module != VMM_KSTAT_MODULE_NAME {
+        return None;
+    }
+    if kstat.ks_name != VM_KSTAT_NAME {
+        return None;
+    }
+    let Data::Named(named) = data else {
+        return None;
+    };
+
+    // Return the instance if the `vm_name` kstat matches our instance UUID.
+    if named.iter().any(|nd| {
+        if nd.name != VM_NAME_KSTAT {
+            return false;
+        }
+        let NamedData::String(name) = &nd.value else {
+            return false;
+        };
+        instance_id == *name
+    }) {
+        return Some(kstat.ks_instance);
+    }
+    None
+}
+
+// Produce `Sample`s for the `VcpuUsage` metric from the relevant kstats.
+fn produce_vcpu_usage<'a>(
+    vm: &'a VirtualMachine,
+    vcpu_stats: impl Iterator<Item = &'a (DateTime<Utc>, Kstat<'a>, Data<'a>)> + 'a,
+) -> Result<Vec<Sample>, Error> {
```

Comment:
IIUC, we'll bail here if we encounter any invalid values in the kstats. what's the rationale for that versus returning a `Vec` that only contains the good samples?

Reply:
It's a good question. I'm sort of operating under the assumption that if fetching one fails, they all will. I believe that's reasonable, based on my experience with the `kstat` library itself, but I could be missing something. I have a similar issue in omicron (https://github.com/oxidecomputer/omicron/issues/4863), where we could actually opt to produce explicitly-missing samples in this case.

I would be happy to make a follow-up issue to do that, if it seems like a good idea.

Reply:
Thanks, I think that's entirely reasonable --- I was just curious about the motivation here. It might be nice to add a comment about this, but it's certainly not a blocker.